rustls/src/client/common.rs

```diff
@@ -1,18 +1,118 @@
+#[allow(unreachable_pub)] // conditional on `cfg(feature = "dangerous_configuration")`
+pub struct ServerVerifyInput<'a> {
+    /// Certificate for the end entity.
+    ///
+    /// Note that none of the certificates have been parsed yet, so it is the responsibility of
+    /// the implementor to handle invalid data. It is recommended that the implementor returns
+    /// [`crate::Error::InvalidCertificateEncoding`] when these cases are encountered.
+    pub end_entity: &'a Certificate,
+    /// Intermediate certificates the client sent along with the end-entity certificate.
+    ///
+    /// It is in the same order that the peer sent them and may be empty.
+    ///
+    /// Note that none of the certificates have been parsed yet, so it is the responsibility of
+    /// the implementor to handle invalid data. It is recommended that the implementor returns
+    /// [`crate::Error::InvalidCertificateEncoding`] when these cases are encountered.
+    pub intermediates: &'a [Certificate],
+    /// The server name the certificate should be valid for.
+    pub server_name: &'a ServerName,
+    /// The time at which the certificate must be valid.
+    pub now: SystemTime,
+    /// The certificate's signature.
+    pub server_signature: &'a DigitallySignedStruct,
+    pub(crate) details: &'a ServerCertDetails,
+    pub(crate) version: ServerVerifyVersionInput<'a>,
+}
+
+#[allow(unreachable_pub)] // conditional on `cfg(feature = "dangerous_configuration")`
+impl<'a> ServerVerifyInput<'a> {
+    /// Construct the signature message that the server should have signed
+    pub fn server_verify_message(&self) -> Vec<u8> {
+        use ServerVerifyVersionInput::*;
+        match self.version {
+            #[cfg(feature = "tls12")]
+            Tls12 { randoms, kx_params } => {
+                // Build up the contents of the signed message.
+                // It's ClientHello.random || ServerHello.random || ServerKeyExchange.params
+                let mut message = Vec::new();
+                message.extend_from_slice(&randoms.client);
+                message.extend_from_slice(&randoms.server);
+                message.extend_from_slice(kx_params);
+                message
+            }
+            Tls13 { transcript } => {
+                // Constructs the signature message specified in section 4.4.3 of RFC8446
+                let handshake_hash = transcript.get_current_hash();
+                construct_tls13_server_verify_message(&handshake_hash)
+            }
+        }
+    }
+
+    /// Signed Certificate Timestamps (SCTs) provided by the server, if any.
+    pub fn scts(&self) -> impl Iterator<Item = &[u8]> {
+        self.details
+            .scts
+            .as_deref()
+            .unwrap_or(&[])
+            .iter()
+            .map(|payload| payload.0.as_slice())
+    }
+
+    /// OCSP response received from the server.
+    #[inline]
+    pub fn ocsp_response(&self) -> &[u8] {
```

Comment:
It's a little surprising that ServerVerifyInput mostly has reference fields that can be accessed directly, but `scts` and `ocsp_response` have to go through an accessor. This seems overly tied to implementation: the fact that `ServerCertDetails` is a `pub(crate)` struct.